File: 315/b.java
import java.io.*;

class main
{
	public static void main(String[] args) throws Exception
	{
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

		String[] options = reader.readLine().split(" ");

		int length = Integer.parseInt(options[0]);
		int queries = Integer.parseInt(options[1]);

		int increment = 0;

		String[] seqAsString = reader.readLine().split(" ");

		int[] seq = new int[length];

		for(int i=0;i<length;i++)
			seq[i] = Integer.parseInt(seqAsString[i]);

		for(int i=0;i<queries;i++)
		{
			String[] query = reader.readLine().split(" ");

			int qMode = Integer.parseInt(query[0]);
			int qData = Integer.parseInt(query[1]);

			if(qMode == 3)
				writer.write(Integer.toString(seq[qData - 1] + increment)+"\n");

			/*
				Save time by leaving the array elements alone and keeping an increment variable.
			 */
			else if(qMode == 2)
				increment += qData;

			else
				seq[qData - 1] = Integer.parseInt(query[2]) - increment;
		}

		writer.flush();
	}
}

File: 315/b.cpp
#include <cstdio>

using namespace std;

int main()
{
	int length,queries,increment = 0;

	scanf("%d%d",&length,&queries);

	int seq[length];

	for(int i=0;i<length;i++)
		scanf("%d",&seq[i]);

	for(int i=0;i<queries;i++)
	{
		int q_type,temp;

		scanf("%d",&q_type);

		if(q_type == 3)
		{
			scanf("%d",&temp);
			printf("%d\n",seq[temp-1]+increment);
		}

		else if(q_type == 2){
			scanf("%d",&temp);
			increment += temp;
		}

		else
		{
			scanf("%d",&temp);
			scanf("%d",&seq[temp-1]);
			seq[temp-1] -= increment;
		}
		
	}

	return 0;
}